src/control/character_controller.rs

```diff
@@ -473,57 +474,41 @@ impl KinematicCharacterController {
         &self,
         hit: &TOI,
         translation_remaining: &Vector<Real>,
-    ) -> (Option<Vector<Real>>, Real) {
-        let vertical_translation_remaining = *self.up * (self.up.dot(translation_remaining));
-        let horizontal_translation_remaining =
-            *translation_remaining - vertical_translation_remaining;
-
-        // The idea behind this `if` statement is as follows:
-        // - If there is any amount of horizontal translations, then the intended
-        //   climb/slide down movement is decided by that translation.
-        // - If there is no horizontal translation, then we only have gravity. In that case,
-        //   we take the vertical movement into account to decide if we need to slide down.
-        let sliding_translation_remaining = if horizontal_translation_remaining != Vector::zeros() {
-            horizontal_translation_remaining
-                - *hit.normal1 * (horizontal_translation_remaining).dot(&hit.normal1)
-        } else {
-            vertical_translation_remaining
-                - *hit.normal1 * (vertical_translation_remaining).dot(&hit.normal1)
-        };
-
-        // Check if there is a slope we can climb.
+        offset: Real,
+    ) -> Option<Vector<Real>> {
+        let [vertical_translation, horizontal_translation] =
+            self.split_into_components(translation_remaining);
+        let [vertical_slope_translation, horizontal_slope_translation] =
+            [vertical_translation, horizontal_translation]
+            .map(|remaining| subtract_hit(remaining, hit, offset));
+        let slope_translation = horizontal_slope_translation + vertical_slope_translation;
```

Comment:
I see, this is the key to fix the vertical wall problem. I’m not sure why I decided to apply that complex `if` statement to determine the `sliding_translation_remaining`.

Reply:
Yep, together with the block prefaced by `// No slopes or stairs ahead; try to move along obstacle.`. I was quite happy to be able to simplify the `if` construct 😄